### PR TITLE
driver: sensor: icp101xx: adds driver and sample

### PR DESCRIPTION
**Pull request by rbuisson-invn**
Adds official Invensense Inc. driver for icp101xx barometric pressure and temprature sensor. Also adds a sample for this driver.
Validated with custom setup based on nrf52dk_nrf52832 + icp10100 daughter board
  Build ok by running: west twister -T samples\sensor\icp101xx -p nrf52dk_nrf52832

### CODE REVIEW

**Comment by rbuisson-invn** [comment]
Hello @MaureenHelm,

I'm new to Zephyr Twister, it looks like Twister decided to build the ICP101XX sample with board: tdk_robokit1, while the sample.yaml specifies nrf52dk_nrf52832 or arduino_zero.
Is there any option that I miss to force Twister build targeted board?

Thanks and regards,
Remi

File: samples/sensor/icp101xx/src/main.c
Comment (kartben) [inline samples/sensor/icp101xx/src/main.c:1]: unless this is to demonstrate sensor-specific API (which it doesn't look like this one has any), please do not introduce a new sample. Folks can use existing samples to interact with the sensor, for ex. `sensor_shell`

**Review by kartben — CHANGES_REQUESTED** [review]

File: drivers/sensor/tdk/icp101xx/EmbUtils/InvBool.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+
+/** @brief    Custom definition for boolean type to avoid compiler disrepenrencies
+ *	@{
+ */
+
+#ifndef _INV_BOOL_H_
+#define _INV_BOOL_H_
+
+typedef int inv_bool_t;
```
Comment (teburd) [inline drivers/sensor/tdk/icp101xx/EmbUtils/InvBool.h:15]: #include <stdbool.h> instead and use bool? Why do we need to redefine bool values

File: drivers/sensor/tdk/icp101xx/EmbUtils/InvError.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ *	@ingroup EmbUtils
+ *	@{
+ */
+
+#ifndef _INV_ERROR_H_
+#define _INV_ERROR_H_
+
+/** @brief Common error code definition
+ */
+enum inv_error {
```
Comment (teburd) [inline drivers/sensor/tdk/icp101xx/EmbUtils/InvError.h:19]: Zephyr errors are typically errno values, e.g. ENOTSUP, EIO

File: drivers/sensor/tdk/icp101xx/EmbUtils/InvExport.h [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * Copyright (c) 2020 TDK Invensense
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#ifndef _INV_IDD_EXPORT_H_
+#define _INV_IDD_EXPORT_H_
+
+#if defined(_WIN32)
```
Comment (teburd) [inline drivers/sensor/tdk/icp101xx/EmbUtils/InvExport.h:11]: When would we be building this for windows as a shared lib

File: drivers/sensor/tdk/icp101xx/Message.h [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+ * Copyright (c) 2024 TDK Invensense
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _INV_MESSAGE_H_
+#define _INV_MESSAGE_H_
+
+/* Disable Message from the driver */
+#define INV_MSG(...)
```
Comment (teburd) [inline drivers/sensor/tdk/icp101xx/Message.h:11]: This is a bit confusing, each driver has a log level adjustable already so this should not be needed

File: drivers/sensor/tdk/icp101xx/Devices/Drivers/ICP101xx/Icp101xx.h [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+#include <stdint.h>
+#include <assert.h>
+#include <string.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define ICP101XX_I2C_ADDR (0x63)
```
Comment (teburd) [inline drivers/sensor/tdk/icp101xx/Devices/Drivers/ICP101xx/Icp101xx.h:31]: I'm not sure I agree in this case that the vendor code is that much of a benefit. The regmap and functionality here to manipulate the device could easily be made to better fit zephyr's sensor drivers already in the tree without a lot of effort.

Please take a look at drivers like the bmi160 and bme680 which are much closer to how I'd expect to see a sensor driver for Zephyr look

File: drivers/sensor/tdk/icp101xx/icp101xx.c [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+#include "icp101xx.h"
+
+LOG_MODULE_REGISTER(ICP101XX, CONFIG_SENSOR_LOG_LEVEL);
+
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 0
+#error "ICP101XX driver enabled without any devices"
+#endif
+
+void inv_icp101xx_sleep_us(int us)
```
Comment (teburd) [inline drivers/sensor/tdk/icp101xx/icp101xx.c:26]: we don't need to redefine k_sleep

File: drivers/sensor/tdk/icp101xx/icp101xx.c [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+		return -ENOTSUP;
+	}
+	rc = inv_icp101xx_enable_sensor(&data->icp_device, 1);
+	/* Compute timeout for the measure */
+	timeout = k_uptime_get() + get_timeout_ms(data->icp_device.measurement_mode);
+	/* Initial sleep waiting the sensor proceeds with the measure */
+	k_sleep(K_MSEC(get_conversion_ms(data->icp_device.measurement_mode)));
+	do {
+		k_sleep(K_USEC(200));
+		rc = inv_icp101xx_get_data(&data->icp_device, &(data->raw_pressure),
```
Comment (teburd) [inline drivers/sensor/tdk/icp101xx/icp101xx.c:140]: this should read the registers and stash them in device data, that's it. This seems to be doing conversions to float

File: drivers/sensor/tdk/icp101xx/icp101xx.c [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+	float pressure;
+	icp101xx_data *data = (icp101xx_data *)dev->data;
+
+	if (!(chan == SENSOR_CHAN_AMBIENT_TEMP || chan == SENSOR_CHAN_PRESS ||
+	      SENSOR_CHAN_ALTITUDE)) {
+		return -ENOTSUP;
+	}
+	/* Zephyr expects kPa while ICP101xx returns Pa */
+	pressure = data->pressure / 1000;
+	if (chan == SENSOR_CHAN_AMBIENT_TEMP) {
```
Comment (teburd) [inline drivers/sensor/tdk/icp101xx/icp101xx.c:159]: Here is where you should take your raw sensor register value and directly convert it to a sensor_value using integer math for each supported channel

File: drivers/sensor/tdk/icp101xx/icp101xx.c [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+	data->icp_device.measurement_mode = cfg->mode;
+
+	/* successful init, return 0 */
+	return 0;
+}
+
+static const struct sensor_driver_api icp101xx_api_funcs = {
+	.sample_fetch = icp101xx_sample_fetch,
+	.channel_get = icp101xx_channel_get,
+	.attr_set = icp101xx_attr_set,
```
Comment (teburd) [inline drivers/sensor/tdk/icp101xx/icp101xx.c:211]: I'd really recommend implementing the submit and decoder APIs as well, as those are the future of Zephyr sensor drivers. In which case for this device on a request to read you would...

* Enable the device to sample
* Start a k_timer
* In the k_timer callback start a i2c read to directly read into a user given buffer
* On i2c read complete you'd mark the read complete

In the decoder you'd implement the ability to then decode the sensor data from its internal register format, stashed in the buffer, to useful fixed point q values.

File: drivers/sensor/tdk/icp101xx/icp101xx.h [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+#include "Devices\Drivers\ICP101xx\Icp101xxSerif.h"
+
+#define DT_DRV_COMPAT invensense_icp101xx
+
+#define ICP101XX_BUS_I2C DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+
+typedef struct {
+	int raw_pressure;
+	int raw_temperature;
+	float pressure;
```
Comment (teburd) [inline drivers/sensor/tdk/icp101xx/icp101xx.h:25]: floats require some special care in Zephyr as floats are *not* always stashed on context swaps, not always supported by the hardware (cortex-m0), and generally not accepted in sensor drivers for those reasons

File: drivers/sensor/tdk/icp101xx/icp101xx.h [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/sensor.h>
+#include "Devices\Drivers\ICP101xx\Icp101xx.h"
+#include "Devices\Drivers\ICP101xx\Icp101xxSerif.h"
+
+#define DT_DRV_COMPAT invensense_icp101xx
+
+#define ICP101XX_BUS_I2C DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+
+typedef struct {
```
Comment (teburd) [inline drivers/sensor/tdk/icp101xx/icp101xx.h:22]: We don't typedef structs in zephyr typically

File: samples/sensor/icp101xx/README.rst [hunk 12]
```diff
@@ -0,0 +1,49 @@
+.. _icp101xx:
+
+ICP101xx: Barometric Pressure and Temperature Sensor
```
Comment (teburd) [inline samples/sensor/icp101xx/README.rst:3]: We have a dht_polling sample that covers a similar case, I don't think we need a device specific sample in this scenario

File: samples/sensor/icp101xx/src/main.c [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+		return NULL;
+	}
+
+	printk("Found device \"%s\", getting sensor data\n", dev->name);
+	return dev;
+}
+
+int main(void)
+{
+	const struct device *dev = get_icp101xx_device();
```
Comment (teburd) [inline samples/sensor/icp101xx/src/main.c:43]: const struct device *dev = DEVICE_DT_GET(DT_ALIAS(press_temp_altitude));

Any board overlay that supplies this devicetree alias then works with this sample. The sample could be renamed to something along those lines. BMP180 from bosch would also fit in this use case and sample, nothing icp101xxx specific should be needed.

**Review by teburd — COMMENTED** [review]
A first pass of this, I will likely need to look it over again

File: drivers/sensor/tdk/icp101xx/EmbUtils/InvBool.h (hunk 1 above)
Comment (rbuisson-invn) [reply drivers/sensor/tdk/icp101xx/EmbUtils/InvBool.h:15]: The goal of his PR is to make sure Zephyr driver for ICP101xx sensors is using the official driver from TDK.
This code is part of this official TDK driver and cannot be modified for Zephyr.

File: drivers/sensor/tdk/icp101xx/EmbUtils/InvExport.h (hunk 3 above)
Comment (rbuisson-invn) [reply drivers/sensor/tdk/icp101xx/EmbUtils/InvExport.h:11]: The goal of his PR is to make sure Zephyr driver for ICP101xx sensors is using the official driver from TDK.
This code is part of this official TDK driver and cannot be modified for Zephyr.

File: drivers/sensor/tdk/icp101xx/Message.h (hunk 4 above)
Comment (rbuisson-invn) [reply drivers/sensor/tdk/icp101xx/Message.h:11]: The goal of his PR is to make sure Zephyr driver for ICP101xx sensors is using the official driver from TDK.
This code is part of this official TDK driver and cannot be modified for Zephyr.

File: drivers/sensor/tdk/icp101xx/Devices/Drivers/ICP101xx/Icp101xx.h (hunk 5 above)
Comment (rbuisson-invn) [reply drivers/sensor/tdk/icp101xx/Devices/Drivers/ICP101xx/Icp101xx.h:31]: The goal of his PR is to make sure Zephyr driver for ICP101xx sensors is using the official driver from TDK.
This code is part of this official TDK driver and cannot be modified for Zephyr.

File: drivers/sensor/tdk/icp101xx/icp101xx.c (hunk 6 above)
Comment (rbuisson-invn) [reply drivers/sensor/tdk/icp101xx/icp101xx.c:26]: inv_icp101xx_sleep_us is a HAL required by the ICP101xx official driver by TDK.

File: drivers/sensor/tdk/icp101xx/icp101xx.c (hunk 7 above)
Comment (rbuisson-invn) [reply drivers/sensor/tdk/icp101xx/icp101xx.c:140]: The TDK official driver for ICP101xx is using float.
I understand that some care must be taken when using float in Zephyr.
Do you confirm that using the k_float_enable/k_float_disable within the driver would make things safer?

File: drivers/sensor/tdk/icp101xx/icp101xx.h (hunk 10 above)
Comment (rbuisson-invn) [reply drivers/sensor/tdk/icp101xx/icp101xx.h:25]: Do you confirm that using the k_float_enable/k_float_disable within the driver would make things safer?
As said in other comments, the official TDK driver for ICP101xx sensor relies on float.

File: drivers/sensor/tdk/icp101xx/icp101xx.h (hunk 11 above)
Comment (rbuisson-invn) [reply drivers/sensor/tdk/icp101xx/icp101xx.h:22]: Agreed, will be changed.

File: samples/sensor/icp101xx/README.rst (hunk 12 above)
Comment (rbuisson-invn) [reply samples/sensor/icp101xx/README.rst:3]: DHT is a sensor from another manufacturer (Aosong Electronics) and is NOT a pressure sensor, this sample can definitely not be used with the TDK ICP101xx pressure sensor.

File: samples/sensor/icp101xx/src/main.c (hunk 13 above)
Comment (rbuisson-invn) [reply samples/sensor/icp101xx/src/main.c:43]: I understand the request, however there are already many Zephyr pressure sensor samples implemented this way (while not having any specific code), such as:
- BME280 from Bosh
- BME680 from Bosh
- DPS310 from Infineon
- LPS22HB from STMicroelectronics

File: samples/sensor/icp101xx/src/main.c
Comment (rbuisson-invn) [reply samples/sensor/icp101xx/src/main.c:1]: I understand the request, however there are already many Zephyr pressure sensor samples implemented this way (while not having any specific code), such as:
- BME280 from Bosh
- BME680 from Bosh
- DPS310 from Infineon
- LPS22HB from STMicroelectronics

File: samples/sensor/icp101xx/src/main.c (hunk 13 above)
Comment (kartben) [reply samples/sensor/icp101xx/src/main.c:43]: I think the point here is: does this sample really provide more than `sensor_shell` sample to an end-user in allowing them to easily exercise their sensor?

Arguably, the only added value is showing how one's DT overlay would look like but this is something that can also easily be done in the documentation of the binding directly if you really want to give users an example of how things could look like ; for example: https://github.com/zephyrproject-rtos/zephyr/blob/553a1c0063d81a6ab088f6632e5803b1e1d43418/dts/bindings/sensor/sensirion%2Cshtcx.yaml#L10-L21)

Comment (kartben) [reply samples/sensor/icp101xx/src/main.c:43]: I think the point here is: does this sample really provide more than `sensor_shell` sample to an end-user in allowing them to easily exercise their sensor?

Arguably, the only added value is showing how one's DT overlay would look like but this is something that can also easily be done in the documentation of the binding directly if you really want to give users an example of how things could look like ; for example: https://github.com/zephyrproject-rtos/zephyr/blob/553a1c0063d81a6ab088f6632e5803b1e1d43418/dts/bindings/sensor/sensirion%2Cshtcx.yaml#L10-L21)

Comment (kartben) [reply samples/sensor/icp101xx/src/main.c:43]: and to be clear, the "generic" samples you're referring to should probably just go away

**Pull request closed without merging**